src/Common/DNSResolver.cpp

```diff
@@ -437,6 +437,20 @@ void DNSResolver::addToNewAddresses(const Poco::Net::IPAddress & address)
     impl->new_addresses.insert({address, consecutive_failures});
 }
 
+std::vector<DNSResolver::CacheEntry> DNSResolver::cacheEntries() const
+{
+    std::vector<DNSResolver::CacheEntry> entries;
+
+    for (const auto & host : impl->known_hosts) {
+        const DB::String &hostname = host.first;
+
+        DB::DNSResolver::IPAddressesPtr lookup_result = impl->cache_host.get(hostname);
+        if (lookup_result)
+            entries.push_back(DNSResolver::CacheEntry(hostname, *lookup_result));
+    }
+    return entries;
```

Comment:
I suppose it's ok implementation for introspection.
However, I think ideally it's better to add a special method to return all the content of cache.
It may be a bit harder, cause `CacheBase` and `ICachePolicy` should be updated. From what I can see, cache policy stores `unordered_map` and iterating it would be better than multiple lookups.
https://github.com/ClickHouse/ClickHouse/blob/62debf42b4d23bc9bd58047e454ac5bd46dc7afe/src/Common/LRUCachePolicy.h#L134

@pykello it's up to you to decide if you like to change it :D

Reply:
@KochetovNicolai you are right that iterating over the internal maps will be more efficient than multiple lookups, however because that internal map seems to be an implementation detail of each cache policy, I would prefer to not directly expose it.

For example, the `Cell` struct used in that unordered_map is defined slightly differently for SLRU & LRU policies.

```cpp
// LRU Cell
   struct Cell
    {
        MappedPtr value;
        size_t size;
        LRUQueueIterator queue_iterator;
    };

// SLRU Cell
    struct Cell
    {
        bool is_protected = false;
        MappedPtr value;
        size_t size;
        SLRUQueueIterator queue_iterator;
    };
```

For that reason, I would ideally add an iterator() method for CachePolicy & each of cache policies would implement it & then CacheBase can expose that iterator.

What do you think?

I am travelling until Dec 22. I can update this after returning back.